pkg/controllers/machine/controller.go

```diff
@@ -0,0 +1,201 @@
+import (
+	"context"
+	"fmt"
+
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/events"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter-core/pkg/termination"
+	"github.com/aws/karpenter-core/pkg/utils/result"
+)
+
+var _ corecontroller.FinalizingTypedController[*v1alpha5.Machine] = (*Controller)(nil)
+
+// Controller is a Machine Controller
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	recorder      events.Recorder
+	terminator    *termination.Terminator
+
+	launch         *Launch
+	registration   *Registration
+	initialization *Initialization
+	liveness       *Liveness
+}
+
+// NewController is a constructor for the Machine Controller
+func NewController(clk clock.Clock, kubeClient client.Client, cloudProvider cloudprovider.CloudProvider,
+	terminator *termination.Terminator, recorder events.Recorder) corecontroller.Controller {
+	return corecontroller.Typed[*v1alpha5.Machine](kubeClient, &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		recorder:      recorder,
+		terminator:    terminator,
+
+		launch:         &Launch{kubeClient: kubeClient, cloudProvider: cloudProvider},
+		registration:   &Registration{kubeClient: kubeClient},
+		initialization: &Initialization{kubeClient: kubeClient},
+		liveness:       &Liveness{clock: clk, kubeClient: kubeClient},
+	})
+}
+
+func (*Controller) Name() string {
+	return "machine"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	// Add the finalizer immediately since we shouldn't launch if we don't yet have the finalizer.
+	// Otherwise, we could leak resources
+	stored := machine.DeepCopy()
+	controllerutil.AddFinalizer(machine, v1alpha5.TerminationFinalizer)
+	if !equality.Semantic.DeepEqual(machine, stored) {
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+
+	stored = machine.DeepCopy()
+	var results []reconcile.Result
+	var errs error
+	for _, reconciler := range []interface {
+		Reconcile(context.Context, *v1alpha5.Machine) (reconcile.Result, error)
```

Comment:
the inline interface is awkward to me, wdyt about just defining it at the top of this file as a `MachineReconciler`?